RATEST-258:Migrate all selenium tests into qaframework from distro

sherrif10 opened the pull request:
https://issues.openmrs.org/browse/RATEST-258

kdaud commented:
@sherrif10 have been testing out things with CI on this subject and realized the following that may be of help;

1. Borrowing a leaf from @mozzy11's  [comment](https://github.com/openmrs/openmrs-contrib-qaframework/pull/201#issuecomment-960667475), all the property values for `login.auto` should be set to `true` in [test-bamboo.properties](https://github.com/openmrs/openmrs-contrib-qaframework/blob/2c6702e124832a5ae716ea61b50db08399919af0/src/test/resources/org/openmrs/uitestframework/test-bamboo.properties#L6), [test.properties](https://github.com/openmrs/openmrs-contrib-qaframework/blob/2c6702e124832a5ae716ea61b50db08399919af0/src/test/resources/org/openmrs/uitestframework/test.properties#L6) and [test-local.properties](https://github.com/openmrs/openmrs-contrib-qaframework/blob/2c6702e124832a5ae716ea61b50db08399919af0/src/test/resources/org/openmrs/uitestframework/test-local.properties#L6)

```js
login.auto=true
```
2. Adding a profile in pom.xml to guide ci while logging into the server
```js
<profiles>
        <profile>
            <id>ci</id>

            <properties>
                <login.username>admin</login.username>
                <login.password>Admin123</login.password>
            </properties>

            <build>
                <plugins>
                    <plugin>
                        <artifactId>maven-failsafe-plugin</artifactId>
                        <configuration>
                            <skipTests>true</skipTests>
                            <groups>org.openmrs.contrib.qaframework.helper.BuildTests</groups>
                            <forkCount>5</forkCount>
                            <reuseForks>true</reuseForks>
                        </configuration>
                    </plugin>
                </plugins>
            </build>
        </profile>
    </profiles>
````
Take care

kdaud commented:
@sherrif10 have been testing out things with CI on this subject and realized the following that may be of help;

1. Borrowing a leaf from @mozzy11's  [comment](https://github.com/openmrs/openmrs-contrib-qaframework/pull/201#issuecomment-960667475), all the property values for `login.auto` should be set to `true` in [test-bamboo.properties](https://github.com/openmrs/openmrs-contrib-qaframework/blob/2c6702e124832a5ae716ea61b50db08399919af0/src/test/resources/org/openmrs/uitestframework/test-bamboo.properties#L6), [test.properties](https://github.com/openmrs/openmrs-contrib-qaframework/blob/2c6702e124832a5ae716ea61b50db08399919af0/src/test/resources/org/openmrs/uitestframework/test.properties#L6) and [test-local.properties](https://github.com/openmrs/openmrs-contrib-qaframework/blob/2c6702e124832a5ae716ea61b50db08399919af0/src/test/resources/org/openmrs/uitestframework/test-local.properties#L6)

```js
login.auto=true
```
2. Adding a profile in pom.xml to guide ci while logging into the server
```js
<profiles>
        <profile>
            <id>ci</id>

            <properties>
                <login.username>admin</login.username>
                <login.password>Admin123</login.password>
            </properties>

            <build>
                <plugins>
                    <plugin>
                        <artifactId>maven-failsafe-plugin</artifactId>
                        <configuration>
                            <skipTests>true</skipTests>
                            <groups>org.openmrs.contrib.qaframework.helper.BuildTests</groups>
                            <forkCount>5</forkCount>
                            <reuseForks>true</reuseForks>
                        </configuration>
                    </plugin>
                </plugins>
            </build>
        </profile>
    </profiles>
````
Take care

sherrif10 commented:
@kdaud  sure .i think i had already added the profilers, Let me update the test.ci.property file. Thanks

kdaud commented:
@sherrif10 checkout the 3rd [observation](https://github.com/openmrs/openmrs-contrib-qaframework/pull/212#issuecomment-961634620) that had left!

kdaud commented:
@sherrif10 is there any reason why the selenium legacy tests are annotated with @Ignore tag?

sherrif10 commented:
Yes please, selenium tests run perfect locally and fail on ci, so am
investing why they are failing

On Sun, Nov 7, 2021, 10:26 AM Kakumirizi Daud ***@***.***>
wrote:

> @sherrif10 <https://github.com/sherrif10> is there any reason why the
> selenium legacy tests are annotated with @ignore
> <https://github.com/ignore> tag?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openmrs/openmrs-contrib-qaframework/pull/212#issuecomment-962564421>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALOKS6D67VJFYZOKJFYBZILUKYS3TANCNFSM5HLKDYMA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

kdaud commented:
did you take a look at the comment [above](https://github.com/openmrs/openmrs-contrib-qaframework/pull/212#issuecomment-961634620)?

kdaud commented:
did you take a look on what I mentioned [above](https://github.com/openmrs/openmrs-contrib-qaframework/pull/212#issuecomment-961634620)?

sherrif10 commented:
> did you take a close look at the comment [above](https://github.com/openmrs/openmrs-contrib-qaframework/pull/212#issuecomment-961634620)?

Yes all the changes you are referring too, i already added them in the pom file

kdaud commented:
@sherrif10 are you sure you looked into all the three ideas mentioned [above](https://github.com/openmrs/openmrs-contrib-qaframework/pull/212#issuecomment-961634620)? Forexample
1. are you sure the `login.auto` value is set to `true`  in test-bamboo.properties and test-local.properties files?
2. are you sure you have added a workflow that will trigger the selenium legacy tests to run on continuous integration?

Have triggered `pull/215` on the repo that you can take a look and have this guy done!

kdaud commented:
any reason for removing the github workflow([.yml file](https://github.com/openmrs/openmrs-contrib-qaframework/pull/212/commits/66dc3fe023516cf2cb008fab3203d29503a39728)) for the selenium legacy tests?

sherrif10 commented:
cc @kdaud , Before i request help from other reviewers , do you mind looking into this test, These tests pass locally when when run as single tests (run as junit tests) however they fail on github actions, they are tests like startvisit which will pass but trigger All FireFox to fail.

sherrif10 commented:
@kdaud  kindly requesting for reviews, We get this resolved.Thanks.

sherrif10 commented:
Hey @kdaud  if you are still concentrating on [this](https://github.com/openmrs/openmrs-contrib-qaframework/pull/217), it would be quite bera if we add more effort on this pr since i already fixed those issues you will be facing, and have this approved and fixed. :smiling_face_with_three_hearts:

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/FormTest.java:
```diff
@@ -0,0 +1,111 @@
+
+/**
+ * Created by sharif
+ */
```
any reason for adding comment?

sherrif10 replied:
Probably for referring author of the class.

sherrif10 commented:
cc @kdaud  Looking forward to see more reviews, This guy is longing to sit in master

kdaud commented on .gitignore:
```diff
@@ -13,3 +13,6 @@ node_modules/
 .DS_Store
+/.apt_generated/
+/.apt_generated_tests/
+/bin/
```
any reason for adding these files?

sherrif10 replied:
Thanks, these files are very usefull while checking out bin folder and instantiating test classes. , scripts for selenium Mere removing of it result in to un wanted changes . So they are very helpful in UI tests.

kdaud replied:
> Mere removing of it result in to un wanted changes 

are you sure we need these files?

sherrif10 replied:
Resolved thanks.

sherrif10 commented:
@kdaud is this ready for merge?

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/AddPatientAppointmentTest.java:
```diff
@@ -0,0 +1,76 @@
+	@Test
+	@Category(BuildTests.class)
+	public void addPatientAppointmentTest() throws Exception {
+		
```
any reason for putting an empty line here?

sherrif10 replied:
Let me remove it quickly thanks.

kdaud commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,33 @@
+      - name: Build qaframework
+        run: |
+          cp -f src/test/resources/org/openmrs/uitestframework/test-local.properties src/test/resources/org/openmrs/uitestframework/test-bamboo.properties src/test/resources/org/openmrs/uitestframework/test.properties
+          mvn clean install -B -DskipTests=false
```
is the workflow `name` smart enough to tell out what tests are being triggered?

kdaud replied:
so is there any reason why you added it?

sherrif10 commented:
Lets not consider refapp3.x refapp badge as a sounding blocker for this pull request because it keeps triggering itself even when its disabled.

sherrif10 commented:
cc @kdaud  waiting for approval , otherwise we waiting for merge.

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/AddNewAllergyTest.java:
```diff
@@ -0,0 +1,61 @@
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+
```
is the empty line here necessary?

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/AddDiagnosisToVisitNoteTest.java:
```diff
@@ -0,0 +1,65 @@
+	private void createTestVisit() {
+		new TestData.TestVisit(patient.uuid, TestData.getAVisitType(), getLocationUuid(homePage)).create();
+	}
+}
```
instead of invoking `createTestVisit()` within the setUp() method, we can rather write its implementation within the @Before method and this will improve code quality. This applies to all other test classes with similar implementation!

sherrif10 replied:
Let us keep things the way they are and we will refactor everything in a new pull request, Am requesting

kdaud replied:
any reason for not doing everything at once and opting to create tickets to fix this?

sherrif10 replied:
Lets keep things the way they were created from distro, then we will later refactor these selenium tests.

kdaud replied:
does it cost any thing if the tests are refactored before the commit is merged?

sherrif10 replied:
Am actually not getting what you are trying to suggest here , do you mind adding more pointers, to my understanding I see no difference.

kdaud replied:
```js
@Before
    public void setup() throws Exception {
        patient = createTestPatient();
        new TestData.TestVisit(patient.uuid, TestData.getAVisitType(), getLocationUuid(homePage)).create();
    }

    @Test
    @Category(BuildTests.class)
    public void addDiagnosisToVisitNoteTest() {
    
    }

    @After
    public void tearDown() throws Exception {
        deletePatient(patient);
    }
}
``

sherrif10 replied:
@kdaud  Thanks for suggestion however i do believe this does the same thing. And again for simplicity and understanding of code workflow. This very method will do us good with clear java approach.

kdaud replied:
having everything defined within setUp() method would be the better approach and this should apply in all other test classes!

sherrif10 replied:
Thanks, i think you are simply complicating things here. We have used this such kind of approach for almost the entire project.Let me be little detailed. 

Defining a method and simply calling it somewhere makes the code more clean and readable probably understanding.

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/AddEditAppointmentBlockTest.java:
```diff
@@ -0,0 +1,72 @@
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+
```
is the empty line here necessary?

sherrif10 replied:
Resolved thanks @kdaud

sherrif10 commented:
@kdaud  kindly approve such that we move to something different.

sherrif10 commented:
Am very sure @dkayiwa  is watching ,@kdaud  Lets finish this up. Let me know incase anything is not clear. Thanks.

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/AddDiagnosisToVisitNoteTest.java:
```diff
@@ -0,0 +1,65 @@
+	private TestData.PatientInfo patient;
+	
+	@Before
+	public void setup() {
```
isn't this supposed to be `setUp()`?

sherrif10 replied:
Resolved all the setUp convention. Thanks and nic catch :+1:

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/AddNewAllergyTest.java:
```diff
@@ -0,0 +1,60 @@
+public class AddNewAllergyTest extends ReferenceApplicationTestBase {
+	
+	private static final String DRUG_NAME = "Aspirin";
+	
```
any reason for adding an empty line here?

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/RecordMergeIssueTest.java:
```diff
@@ -0,0 +1,93 @@
+import org.openmrs.contrib.qaframework.page.HomePage;
+import org.openmrs.contrib.qaframework.page.RegistrationPage;
+
+
```
isn't one empty line enough to separate imports and  class declaration?

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/TransferToWardServiceTest.java:
```diff
@@ -0,0 +1,59 @@
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ * <p>
```
do we need `<p>`?

sherrif10 commented:
This thing takes almost 2 hrs to get sorted out and all the workflows to run and finish that's why it needs some patience.Again  RefApp 3.x  patient registration makes life hard a bit, It keeps hanging and hence breaking the build , however good news is that its under development and it have no harm on the system since its disable.

sherrif10 commented:
@kdaud  Have resolved all your comments. Any more reviews.

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/AddDeleteLocationTest.java:
```diff
@@ -0,0 +1,61 @@
+	
+	@Test
+	@Category(BuildTests.class)
+	public void addLocationTest() {
```
isn't this supposed to be `addDeleteLocationTest()`?

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/VisitNoteTest.java:
```diff
@@ -0,0 +1,94 @@
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ * <p>
```
do we need `<p>`?

sherrif10 replied:
Ohh it was an oversight. Resolved.Thanks

sherrif10 commented:
cc @kdaud .Thanks for amazing reviews.

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/AddRolesToUserTest.java:
```diff
@@ -0,0 +1,86 @@
+		ManageUserPage manageUserPage = administrationPage.clickOnManageUsers();
+		manageUserPage.removeUser(NURSE_USERNAME);
+	}
+}
```
why adding a test that is ignored to run with ci? AddRolesToUserTest seems to have not been constructed well during its development and I think better not to add it since we have [UserAccount](https://github.com/openmrs/openmrs-contrib-qaframework/blob/master/src/test/java/org/openmrs/contrib/qaframework/automation/UserAccountSteps.java) workflow test that covers adding roles to users, checkout how roles are added to [users ](https://github.com/openmrs/openmrs-contrib-qaframework/blob/6b5c48d722e09c0ab46f1dfd2a160d244ec2642d/src/test/java/org/openmrs/contrib/qaframework/automation/UserAccountSteps.java#L85-L88)

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/AddRolesToUserTest.java:
```diff
@@ -0,0 +1,86 @@
+import org.openmrs.contrib.qaframework.page.ManageUserPage;
+
+@Ignore
+public class AddRolesToUserTest extends ReferenceApplicationTestBase {
```
why adding a test that is ignored to run with ci? AddRolesToUserTest seems to have not been constructed well during its development and I think better not to add it since we have [UserAccount](https://github.com/openmrs/openmrs-contrib-qaframework/blob/master/src/test/java/org/openmrs/contrib/qaframework/automation/UserAccountSteps.java) workflow test that covers adding roles to users, checkout how roles are added to [users ](https://github.com/openmrs/openmrs-contrib-qaframework/blob/6b5c48d722e09c0ab46f1dfd2a160d244ec2642d/src/test/java/org/openmrs/contrib/qaframework/automation/UserAccountSteps.java#L85-L88)

sherrif10 commented:
This ticket is still valid here   https://github.com/openmrs/openmrs-distro-referenceapplication/pull/595  , so i think we  can give the person working on it finish it.

kdaud commented:
have you seen the reason why we would not add it?

sherrif10 commented:
> have you seen the reason [above](https://github.com/openmrs/openmrs-contrib-qaframework/pull/212#discussion_r747226883) why we would not add it?

These actually totally different test units, [AddRolesToUsers ](https://github.com/openmrs/openmrs-contrib-qaframework/blob/6b5c48d722e09c0ab46f1dfd2a160d244ec2642d/src/test/java/org/openmrs/contrib/qaframework/automation/UserAccountSteps.java#L85-L88) add roles to users  from the system , then userAccountSteps  populates users accounts rather  not roles.

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/ManageProviderSchedulesTest.java:
```diff
@@ -0,0 +1,62 @@
+ * Other possibilities are Rules or extending TestSetup. The latter possibilities might be applied
+ * in future if the test code become hard to maintain/read
+ */
+
```
any reason for adding empty line here?

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/UsingBackButtonInMergePatientTest.java:
```diff
@@ -0,0 +1,86 @@
+package org.openmrs.contrib.qaframework.test;
```
why not add header licence on this test?

sherrif10 commented:
Resolved thanks.

sherrif10 commented:
Resolved thanks.

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/ContactInfoTest.java:
```diff
@@ -0,0 +1,78 @@
+		registrationEditSectionPage.enterPhoneNumber("aaaaaaaaa");
+		registrationEditSectionPage.clickOnConfirmEdit();
+		assertTrue(registrationEditSectionPage.getInvalidPhoneNumberNotification()
+		        .contains("Must be a valid phone number (with +, -, numbers or parentheses)"));
```
can't this supposed to be on the same line?

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/EditDemographicTest.java:
```diff
@@ -0,0 +1,62 @@
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ * <p>
```
do we need `<p>`? if not, cross check through all the files with header license and do the needful

sherrif10 commented:
Resolved.

kdaud commented:
@sherrif10 have you taken a look at the build failures?  https://github.com/openmrs/openmrs-contrib-qaframework/runs/4175799088?check_suite_focus=true

sherrif10 commented:
Hey @kdaud lets merge this.

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/RegisterPatientTest.java:
```diff
@@ -0,0 +1,63 @@
+		        + patient.country + ", " + patient.postalCode;
+		
+		assertThat(registrationPage.getNameInConfirmationPage(),
+		    containsString(patient.givenName + ", " + patient.familyName));
```
can't this be on the same line?

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/RegisterPatientTest.java:
```diff
@@ -0,0 +1,63 @@
+		    containsString(patient.givenName + ", " + patient.familyName));
+		assertThat(registrationPage.getGenderInConfirmationPage(), containsString(patient.gender));
+		assertThat(registrationPage.getBirthdateInConfirmationPage(),
+		    containsString(patient.birthDay + ", " + patient.birthMonth + ", " + patient.birthYear));
```
even this here?

sherrif10 commented:
Resolved thanks.

sherrif10 commented:
This build is too large and its hectic for the builds to successfully pass, Its giving me hard time to keep building single build per build. Am even scared that this might slow the time build it takes. :slightly_smiling_face:

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/RecordMergeIssueTest.java:
```diff
@@ -0,0 +1,92 @@
+	@Category(BuildTests.class)
+	public void recordMergeIssueTest() throws Exception {
+		homePage.goToRegisterPatientApp().waitForPage();
+		//       Register first patient
```
isn't this comment supposed to be

```js
// Register first patient
```
?

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/VisitNoteTest.java:
```diff
@@ -0,0 +1,94 @@
+		editVisitNotePage.addNote("This is a new note");
+		patientDashboardPage = visitNotePage.save();
+		
+		//Deleting visit Note.....
```
any reason for putting 4-dots on the comment?

sherrif10 commented:
Resolved thanks.

sherrif10 commented:
@kdaud  Addressed your comments.Let me know if there is anything more to rectify.Otherwise thanks for harming to clean code.

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/UsingBackButtonInMergePatientTest.java:
```diff
@@ -0,0 +1,95 @@
+/**
+ * Created by sharif, kdaudi
+ */
+
```
what is the purpose of adding the comment above?

sherrif10 commented:
Resolved thanks.

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/page/ReportHistoryPage.java:
```diff
@@ -11,20 +11,22 @@
-
+	
 	public RenderDefaultReportPage clickOnViewLink() {
+		waiter.until(ExpectedConditions.elementToBeClickable(VIEW_REPORT));
```
any reason for adding a waiter call here?

sherrif10 commented:
@kdaud , The reason is it thrown an exception element not clickable. The element takes long to be clickable, Adding a waiter element solves the problem.

kdaud commented:
> The reason is it thrown an exception element not clickable

which test is failing?

sherrif10 commented:
> which test is failing?

Reportsteps here https://github.com/openmrs/openmrs-contrib-qaframework/blob/master/src/test/java/org/openmrs/contrib/qaframework/automation/ReportSteps.java

kdaud commented:
is RefApp 2.x Reports workflow test failing with ci?

sherrif10 commented:
> is RefApp 2.x Reports workflow test failing with ci?
Its is not failing but initially we needed to add a waiting element to wait for this element.The ideal reason is , Report renderer takes a lot of time especially when the query is too big, Suppose the defined or quered report is too big. The element will take a lot of time for the reportrenderer to load all the reports and hence this will throw element not clickable exception. Thats the main reason am  adding the waiter here

Another thing is, These reports are not general. so they will be generated having different figures.

kdaud commented:
waiter call(`waiter.until(ExpectedConditions.elementToBeClickable()`); applies when an element takes long to be rendered on the page, does `VIEW_REPORT` element applicable?

sherrif10 commented:
> waiter call(`waiter.until(ExpectedConditions.elementToBeClickable()`) applies when an element takes long to be rendered on the page, does `VIEW_REPORT` element applicable?

@kdaud  Yes its also applicable here.

sherrif10 commented:
Again as you review this, Take in mind that these tests run in parallel. When you trigger any run command forexample npmRunRegistration workflow. All the selenium tests will be triggered as well, At some point this will reduce the strength  the server is running. Then when it reaches on the point of running reports, this is where lagging comes in.thats why such cases of adding waiting element holds.

kdaud commented:
> Then when it reaches on the point of running reports, this is where lagging comes in

can you provide logs to this?

kdaud commented:
> When you trigger any run command forexample npmRunRegistration workflow. All the selenium tests will be triggered as well,

isn't this a poor design?

sherrif10 commented:
> can you provide logs to this

> > Then when it reaches on the point of running reports, this is where lagging comes in
> 
> can you provide logs to this?

sherrif10 commented:
> isn't this a poor design?

this is not poor design by the fact that its the same convention it followed even when tests where running in distro. There is no difference than the format. These will be tracked via all firefox workflow

sherrif10 commented:
But i hope you clearly understand my point of adding a waiter element.Before the driver clickOn ReportView. there is lag in the background when reports are being processed especially when a report is too large. This leads to clickable exception.

kdaud commented:
RefApp 2.x Report test is passing well with ci so there is no need of adding waiter call to VIEW_REPORT, the element doesn't take long to load on the page.

sherrif10 commented:
Since you havent understand my point above about the server taking long due to these changes am introducing with selenium tests,Am going with your suggestion.

sherrif10 commented:
@kdaud  Resolved . let me know if you have more reviews or recommendations please

sherrif10 commented:
Am sure @dkayiwa  is watching and we cant wait for his final reviews or probably merging this guy.

sherrif10 commented:
cc @kdaud  looking foward for more reveiws or approval. Then we invite @dkayiwa ,we seem to have spent too much time on one thing.

kdaud commented:
@sherrif10 as mentioned some time back its a good idea to tag @dkayiwa when the commit has been approved, I see no reason of you tagging him when the PR is still under review.

sherrif10 commented:
is there any good reason why this commit is not approved ?

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/NamePatientAccentedLetterTest.java:
```diff
@@ -0,0 +1,58 @@
+import org.openmrs.contrib.qaframework.page.ClinicianFacingPatientDashboardPage;
+import org.openmrs.contrib.qaframework.page.RegistrationPage;
+
+import io.cucumber.java.After;
```
isn't this supposed to be a junit library?

dkayiwa commented:
> @sherrif10 as mentioned some time back its a good idea to tag @dkayiwa when the commit has been approved, I see no reason of you tagging him when the PR is still under review unless otherwise.

Thank you so much @kdaud for reminding him! 👍

sherrif10 commented:
I did that as reminder since you took time without reviewing and the rule is to wait for your approval until this gets merged. so it was a reminder.

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/test/DeleteAppointmentBlockTest.java:
```diff
@@ -0,0 +1,64 @@
+	@Test
+	@Category(BuildTests.class)
+	public void deleteAppointmentBlockTest() throws Exception {
+		
```
any reason for adding an empty line here?

sherrif10 commented:
Resolved thanks.

kdaud commented:
> these test run parallel and independently when run locally but using the same configuration for running
All Firefox Tests / build.

how can one run all the selenium legacy tests locally?

sherrif10 commented:
simply clone this module and run each test individually as  as junit test  it was in distro.

kdaud commented:
is there any reason why the selenium legacy tests are using the All Firefox Tests configuration and not having its own workflow configuration?

sherrif10 commented:
The reason is because there is no difference than configuration from all
Firefox, since these are triggered by junit test by default allFirefox
configuration is being triggered, ontop of that, this config is supposed to
trigger qa server of which all Firefox config triggers, that's why I see no
reason of adding another config for selenium tests since the behind scenes
is trigger the ci

On Sun, Nov 14, 2021, 4:05 PM Kakumirizi Daud ***@***.***>
wrote:

> is there any reason why the selenium legacy tests are using the All
> Firefox Tests configuration and not having its own workflow configuration?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openmrs/openmrs-contrib-qaframework/pull/212#issuecomment-968287520>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALOKS6FPOZXQL6JNDEDWMFTUL6X2HANCNFSM5HLKDYMA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

kdaud commented:
Have been testing out this PR changes and here is the observation.

When any of the existing workflow tests(refapp2VitalsAndTriaging,  refapp2ClinicalVisit, refapp2Inpatient, refapp2Report, etc)  is triggered to run, all the selenium legacy tests are triggered to run on top of each of them(selenium legacy tests run before the intended test).  This simply means all the selenium legacy tests are running simultaneously with each of the workflow tests on any build plan. 

If the issue is not addressed, here are the outcomes
1. the time taken to run each existing workflow test is increased since each of them has to first run all the 58 selenium legacy tests and any failure in any of the selenium tests during the run will  fail the workflow.
2. `Pain to contributors`. Use case: Imagine some one is developing a new workflow test and is simulation the test locally, the contributor will have to wait for all the selenium legacy  tests to run then the run for the intended workflow comes last.

This need to be addressed, and my recommendation is in the last paragraph of the comment [above](https://github.com/openmrs/openmrs-contrib-qaframework/pull/212#issuecomment-966996576).

sherrif10 commented:
What you tested out , I already did that like I said above there is a big difference from the former configuration we had in distro. And am sure we should except something like this if we decided to make things in typical selenium, 
In a selenium config, we used to checkout uitest sub module and the tests used to run smoothly and independently, that's quite different way this case, all the tests when once triggered will be triggered, feel free to share any configuration that will check out the package tests reside

We should put on mind that all the tests will be first triggered using test runner that is RunTest class, I remember to have explained this before, but since tests are first triggered by default in test runner, that's why we are seeing  all works triggering selenium tests

One way we can try it out is have ng a sub module  that will handle these tests then make a call in a config by checkout the submodule but as long as u a depending on one particular module feel the pain of the server taking long time to load these artifacts

sherrif10 commented:
@kdaud  am open if you recommend a config that might work for the moment but remember we removed a configuration that almost doing same thing.To make things little easy and move on. Let's handle things in particular if the servers and config are up and running, that will simply tell us that selenium tests are passing to.this should be seen on Firefox badge

sherrif10 commented:
That's the pain when it comes with huge runs while this pull request was getting huge and huge but I can assure you that there is no big pain, the tests still have to run when things get bigger and bigger. So we might bare them Again I do believe there is no problem to this approach as long as the tests run independently when run as junit test. Will track things locally using all Firefox built in config

sherrif10 commented:
Is that explanation sufficient.

kdaud commented:
With the current set up for selenium legacy tests in this commit, they are running simultaneously with each workflow tests(this is a poor design). `Use case:` When you trigger refapp2Report workflow test by running `npm run refapp2Report`, all the 58 selenium legacy tests will have to run first before refapp2Report workflow.

Making the selenium legacy tests run independently off the existing workflow tests is the better way!

sherrif10 commented:
That's the exact thing I explained above please. Cross check the explanation above for more enhancement of these tests.

sherrif10 commented:
For these selenium tests to run in their own single workflow , we need to either have a sub module that will be checkout in a config.yml or change everything to a new bdd work which I think was rejected latter, I would think they go with this approach because it also work as expected.

kdaud commented:
why should selenium legacy tests(58 tests) run on top of each of the BDD-tests on every build plan?

sherrif10 commented:
It's because test runner class first trigger all tests without defined
annotations before it checks out the one defined in the package. Especially
the package with glue tag.

On Sun, Nov 14, 2021, 6:06 PM Kakumirizi Daud ***@***.***>
wrote:

> why should selenium legacy tests(58 tests) run on top of each of the
> BDD-tests on every build plan?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openmrs/openmrs-contrib-qaframework/pull/212#issuecomment-968307764>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALOKS6DCYPCZMUOE25DLW5DUL7F6DANCNFSM5HLKDYMA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

kdaud commented:
isn't this a poor design?

sherrif10 commented:
It's not a  really poor design, we shall refactor everything in future but
we can for the moment leave things the way they are now.

On Sun, Nov 14, 2021, 9:54 PM Kakumirizi Daud ***@***.***>
wrote:

> isn't this a poor design?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openmrs/openmrs-contrib-qaframework/pull/212#issuecomment-968344566>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALOKS6ABBPH4LAGDT7ZCVQ3UMAAXRANCNFSM5HLKDYMA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

sherrif10 commented:
Leaving things the way they are for now is the ideal point, Personally i think we decided to do this because of limiting time we had otherwise we would had changed everything into New bdd Approach. So we can leave selenium tests they way they work, However am open if you have more recommendations and suggestions, But personally i would go with this.

kdaud commented:
I do not see any reason why selenium legacy tests should run on top of each BDD tests for every build plan. 

Here are thoughts;
1. why are the selenium legacy tests not having their own workflow that triggers them to run independently?
2. having the tests run independently would be the better way than having them run on top of each BDD tests for every build plan.

sherrif10 commented:
Thanks @kdaud  do you mind sharing your suggestion of how this can be achieved.That would help us a lot and save us time because all the explanations i made  above refers to different approaches and am sure you had a look at them.

kdaud commented:
> do you mind sharing your suggestion of how this can be achieved.

@sherrif10 have you taken a look at the thoughts above?

kdaud commented:
> do you mind sharing your suggestion of how this can be achieved.

@sherrif10 have you taken a look at the thoughts above?

sherrif10 commented:
> have you taken a look at the thoughts above?

@kdaud  which thoughts are those ?

kdaud commented:
> The tests run independently locally

@sherrif10 on your branch, trigger `npm run refapp2Report` or any ref app 2.x workflow test to figure out what am addressing here!

sherrif10 commented:
@kdaud  #**For these selenium tests to run in their own single workflow , we need to either have a sub module that will be checkout in a config.yml or change everything to a new bdd work which I think was rejected latter, I would think we go with this approach because it also work as expected.** did you see this response above.

kdaud commented:
> For these selenium tests to run in their own single workflow , we need to either have a sub module that will be checkout in a config.yml

have you given a short on this and works out well? 

what we need to ensure is that selenium legacy tests do not run on top of the BDD tests. did you see the behavior of the current set up as per this [comment](https://github.com/openmrs/openmrs-contrib-qaframework/pull/212#issuecomment-968516189)?

sherrif10 commented:
I seem to have addressed that comment above. Feel free to cross check above the comments.

kdaud commented:
what happens when you trigger `npm run refapp2Reports`?

sherrif10 commented:
> what happens when you trigger `npm run refapp2Reports`?

The tests run with all selenium tests.

kdaud commented:
> The tests run with all selenium tests.

The BDD tests should run simultaneously with selenium legacy tests. Work around ensuring when the BDD workflows tests do not trigger selenium tests to run on top of them.

sherrif10 commented:
> > The tests run with all selenium tests.
> 
> The BDD tests should **not** run simultaneously with selenium legacy tests. Work around ensuring the BDD workflow tests do not trigger selenium tests to run on top of them.

Did you take a look at this comment [above](https://github.com/openmrs/openmrs-contrib-qaframework/pull/212#issuecomment-968296079), do you mind coming up with a significant workflow that will be used here because all things you are talking about have been tried out, i seem to have lost any option. try to help me with a configuration that might be used here thanks alot.

sherrif10 commented:
I even don't know why we are complicating things, this was an initial migration of selenium tests and we would handle other things in the future since we have limited time. we need to do other things, this is taking long for nothing I think. Then we can handle other things later. refractory is a must unless you haven't felt.

sherrif10 commented:
Let me see if i can come up with submodule for that .

sherrif10 commented:
@kdaud  how is the status of this commit. did you consent with my changes and free to approve them then we later refactor everything or you think this is not worth it.

kdaud commented:
@sherrif10 as long as the selenium legacy tests do not run simultaneously with BDD tests but rather independently. Then the tests will be safe to enter master, checkout the two comments [1](https://github.com/openmrs/openmrs-contrib-qaframework/pull/212#issuecomment-968305653) and [2](https://github.com/openmrs/openmrs-contrib-qaframework/pull/212#issuecomment-968506139)!

sherrif10 commented:
Am adding that as separate submodule  but like i said above this is kinda time-wasting since there is going to be possibility of cleaning all selenium tests into new bdd workflow. Otherwise there is no any other option for doing this.

sherrif10 commented:
Have simply separated the tests into a folder and these tests are running independently without interfering with new BDD  test.

kdaud commented:
which workflow is triggering the selenium legacy tests while running ci?

sherrif10 commented:
@kdaud  hold on i come up with submodule for this workflow.

sherrif10 commented:
This is wrong implementation, Am reverting everything to the original files, Why am i saying like that, we wont create any submodule for that. This is leading me to have distro again as sub module on to this module where by all effort was making the module standalone without depending on each module. So am reverting this to the original work. This wont work at all.

sherrif10 commented:
Let me explain a bit here

1. if you create a separate folder like i did  before tests will not run locally, the idea is to be able to have access to pom files and other dependencies.
2. Another pain point is we would love to keep things simple without having again to depend back on distro
3. Any configuration that will trigger the selenium tests must have a folder that checkout the test.ci.properties where by these files must be in that same folder, This make things more widen
4. selenium tests will get triggered immediately you add them in qaframework module since test runner class first gets instantiated.

sherrif10 commented:
@kdaud  if you are still not agreeing with me on keeping selenium tests run against each run  as it have been , or Make these into new bdd approach, am open to see your sounding commit that works for us.Thanks so much. Otherwise have tried all the options but they end up bringing me back to the old option. Thanks.

sherrif10 commented:
@kdaud  Let me know if you got a cool idea of how to go about this. Thanks.

sherrif10 commented:
@kdaud  Are you having some thing different that can improve this workflow ? or we go with these changes ?

kdaud commented:
@sherrif10 if we can ensure the selenium legacy tests do not run on top of BDD tests then would be sweet. You have done great work with this PR and I acknowledge your effort.

Going with the current setting of the selenium tests is going to cause pain to our contributors working on the new workflows since on every build plan for their respective new tests, all the selenium legacy tests will be running on top of the BDD test and a failure in any of the 58 tests during the run will fail the workflow and perhaps the build time is increased by default.

Have also noticed during CI that some BDD workflows have been failing their build as a result of some selenium legacy tests failing their respective run.

kdaud commented:
Have been taken up by some other tasks and haven't  looked into how this can be achieved!

sherrif10 commented:
> Have also noticed during CI that some BDD workflows have been failing their build as a result of some selenium legacy tests failing their respective run.

Havent yet noticed this, Usually this can easily be tracked easily via the logs.

kdaud commented:
Have been watching how the BDD workflow tests were building on some push you were making on this pull request and I have been observing some of them failing as a result of some selenium legacy tests build failure and often observed rebuilding the tests to have them pass.

I also noticed the BDD workflow tests are taking long to complete the build plan as a result of running simultaneously with selenium legacy tests.

sherrif10 commented:
Thats true , the only way i see and that way could solve this problem is change everything into new bdd approach of which we are overriding.

sherrif10 commented:
@kdaud  we need your response on slack, i had left a comment to address @dkayiwa , so feel free to check on qa slack channel about this approach.
